### PR TITLE
DrivAerML: 4L/768d ultra-wide — more capacity via width scaling

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current DrivAerML best uses 4L/512d/8H achieving 4.619%. We hypothesize that scaling width from 512d to 768d (50% more parameters per layer) at the same depth could push the result lower. Width scaling tends to be more VRAM-efficient than depth scaling for transformer-style architectures, and the 96GB GPUs have ample headroom to accommodate the larger model. We pair this with gradient clipping (gc=1.0) which has proven beneficial across AirfRANS and TandemFoil experiments.

We sweep across 3 LRs (5e-4, 4e-4, 6e-4) and 2 T_max values (30, 10) across all 8 GPUs to find the optimal training regime for the wider architecture.

## Instructions

Run the following 8 experiments across all 8 GPUs. Use `--wandb-group ultra-wide-768d` so all runs are grouped in W&B.

**GPU 0 — 4L/768d, lr=5e-4, gc=1.0, T_max=30 (primary)**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --grad-clip 1.0 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 768 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group ultra-wide-768d \
  --wandb-name "4L-768d-lr5e4-gc1.0-tmax30"
```

**GPU 1 — 4L/768d, lr=5e-4, gc=1.0, T_max=10**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 5e-4 --cosine-t-max 10 \
  --grad-clip 1.0 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 768 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group ultra-wide-768d \
  --wandb-name "4L-768d-lr5e4-gc1.0-tmax10"
```

**GPU 2 — 4L/768d, lr=4e-4, gc=1.0, T_max=30**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=2 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 4e-4 --cosine-t-max 30 \
  --grad-clip 1.0 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 768 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group ultra-wide-768d \
  --wandb-name "4L-768d-lr4e4-gc1.0-tmax30"
```

**GPU 3 — 4L/768d, lr=4e-4, gc=1.0, T_max=10**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=3 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 4e-4 --cosine-t-max 10 \
  --grad-clip 1.0 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 768 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group ultra-wide-768d \
  --wandb-name "4L-768d-lr4e4-gc1.0-tmax10"
```

**GPU 4 — 4L/768d, lr=6e-4, gc=1.0, T_max=30**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=4 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 6e-4 --cosine-t-max 30 \
  --grad-clip 1.0 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 768 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group ultra-wide-768d \
  --wandb-name "4L-768d-lr6e4-gc1.0-tmax30"
```

**GPU 5 — 4L/768d, lr=6e-4, gc=1.0, T_max=10**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=5 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 6e-4 --cosine-t-max 10 \
  --grad-clip 1.0 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 768 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group ultra-wide-768d \
  --wandb-name "4L-768d-lr6e4-gc1.0-tmax10"
```

**GPU 6 — 4L/768d, lr=3e-4, gc=1.0, T_max=30 (conservative LR for large model)**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=6 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 3e-4 --cosine-t-max 30 \
  --grad-clip 1.0 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 768 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group ultra-wide-768d \
  --wandb-name "4L-768d-lr3e4-gc1.0-tmax30"
```

**GPU 7 — 4L/768d, lr=3e-4, gc=1.0, T_max=10**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=7 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 3e-4 --cosine-t-max 10 \
  --grad-clip 1.0 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 768 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group ultra-wide-768d \
  --wandb-name "4L-768d-lr3e4-gc1.0-tmax10"
```

Report the best `val_primary/surface_rel_l2_pct` for each GPU run along with the epoch it was achieved. Include the W&B run URL for each run. Note any OOM issues — if 768d doesn't fit with batch-size 1, try reducing `--drivaerml-train-surface-points` to 40000 or `--drivaerml-eval-surface-points` to 40000.

## Baseline

- **Dataset**: DrivAerML
- **Primary metric**: `val_primary/surface_rel_l2_pct` (lower is better)
- **Current best**: 4.619% — PR #2691 (frieren), epoch 256/267
- **Config**: 4L/512d/8H, AdamW lr=5e-4, T_max=30, no gc, no wd
- **External target**: <3.71%
- **Baseline W&B**: [PR #2691](https://github.com/wandb/senpai/pull/2691)
- **Reproduce command**:
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT guts:

## Results

### Summary Table — 4L/768d Ultra-Wide DrivAerML Sweep

| GPU | LR | T_max | Best `val_primary/surface_rel_l2_pct` | Best Epoch | Final Val% | Total Epochs | Test% | W&B Run |
|---|---|---|---|---|---|---|---|---|
| GPU0 | 5e-4 | 30 | **6.370%** | 130 | 6.370% | 130 | 7.255% | [8yzqnyb3](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/8yzqnyb3) |
| GPU7 | 3e-4 | 10 | **6.457%** | 130 | 6.457% | 130 | 7.307% | [854cswwc](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/854cswwc) |
| GPU2 | 4e-4 | 30 | 7.030% | 125 | 8.959% | 127 | 9.017% | [pk22umzq](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/pk22umzq) |
| GPU3 | 4e-4 | 10 | 7.116% | 124 | 7.116% | 124 | 8.123% | [vg3g50hx](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/vg3g50hx) |
| GPU6 | 3e-4 | 30 | 7.138% | 112 | 8.250% | 120 | 9.163% | [6eadwd5o](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/6eadwd5o) |
| GPU1 | 5e-4 | 10 | 9.349% | 78 | 78.364% | 131 | 79.455% | [vzcvq7bg](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/vzcvq7bg) |
| GPU4 | 6e-4 | 30 | 10.079% | 65 | 91.253% | 125 | 92.936% | [4kiqzwnn](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/4kiqzwnn) |
| GPU5 | 6e-4 | 10 | 25.829% | 21 | 50.066% | 125 | 50.732% | [egndjumc](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/egndjumc) |

**Baseline: 4.619%** (4L/512d/8H, PR #2691, epoch 256/267) | **External target: <3.71%**

### Comparison to Baseline

- **Best result:** 6.370% (GPU0, lr=5e-4, T_max=30) — **37.9% worse** than baseline 4.619%
- **Second best:** 6.457% (GPU7, lr=3e-4, T_max=10) — **39.8% worse** than baseline
- **None of the 8 runs beat the baseline.**

### Commands Used

All 8 runs used the exact commands from the PR instructions with `--agent guts` added. Example (GPU0):
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --grad-clip 1.0 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 768 --model-heads 8 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --agent guts --wandb-group ultra-wide-768d \
  --wandb-name "guts/4L-768d-lr5e4-gc1.0-tmax30"
```

### Peak Memory
~16.4 GB per GPU (out of 97.8 GB available). No OOM issues.

### What Happened

**Width scaling from 512d→768d did not improve DrivAerML performance.** The hypothesis that more capacity via width would push the metric lower was falsified at this training budget.

Key observations:
1. **Throughput penalty is significant:** 768d gets ~130 epochs vs 267 for 512d in the same 360-min budget (~2.75 min/epoch vs ~0.67 min/epoch). The 512d model sees 2x more training iterations.
2. **The throughput-capacity tradeoff strongly favors 512d:** Despite 50% more parameters per layer, 768d converges to a much worse minimum because it can't train long enough. This mirrors the earlier finding that 4L/320d (5.027% at 257 epochs) was competitive with 4L/512d (4.619% at 267 epochs) — throughput matters enormously on DrivAerML.
3. **lr=6e-4 is catastrophically unstable:** Both GPU4 (T_max=30) and GPU5 (T_max=10) diverged badly. The larger model amplifies gradient instability at high LR.
4. **T_max=10 causes late-stage divergence at lr=5e-4:** GPU1 found 9.35% at epoch 78 but then exploded to 78%. T_max=30 is significantly safer for the 768d architecture.
5. **gc=1.0 may be suboptimal for 768d:** The baseline 512d used no gradient clipping. gc=1.0 was added here based on AirfRANS/TandemFoil findings, but it may over-constrain the larger model's optimization landscape.
6. **Best two runs (GPU0, GPU7) both ended at epoch 130 at their best value** — suggesting they hadn't converged yet, but the throughput penalty prevents reaching enough epochs.

### Suggested Follow-ups

1. **The 512d sweet spot is confirmed** — width scaling beyond 512d is counterproductive at this training budget. The 4L/512d model gets the best balance of capacity and throughput.
2. **Try gc=1.0 on the existing 4L/512d** — the baseline didn't use gc, but gc has been beneficial everywhere else. A direct comparison at 512d would be informative.
3. **5L/512d depth scaling** at the proven lr=5e-4/T_max=30 config could be more productive than width scaling.
4. **768d might only work with much longer training budgets** (>1000 epochs) — the convergence curves suggest these models need more iterations to realize their capacity advantage.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2890 because Dead end — 4L/768d ultra-wide is significantly worse than our current 4L/512d baseline. Best run (lr=5e-4, T_max=30) achieved 6.370% vs our current best of 3.997%, a 59% regression. Width scaling beyond 512d appears to hurt DrivAerML — the model likely cannot fit larger representations meaningfully at this data scale, and the extra parameters are adding noise rather than capacity. 

Key findings: lr=5e-4 remains the best LR; lr=6e-4 diverged catastrophically even with gc=1.0; lr=4e-4 and lr=3e-4 also underperformed. The consistent failure across all LRs rules out a hyperparameter fix — 768d is fundamentally worse than 512d for DrivAerML. The current 4L/512d/8H golden config (3.997% at ep467) remains the DrivAerML architecture optimum. Future DrivAerML experiments should focus on different levers: optimizer variants (Lion?), learning rate schedule shape (cosine T_max tuning beyond ep467), or loss reformulation rather than architecture scaling..
